### PR TITLE
feat(medusa): Add ProductVariantInventoryService

### PR DESCRIPTION
**Pull request by pKorsholm**
**What**
- Add services: 
  - `sales-channel-inventory`
  - `sales-channel-location`
  - `product-variant-inventory`
- remove `inventoryService`
- adjust remaining services to match

picks out the core parts of #2459 
Fixes CORE-914, CORE-912

### CODE REVIEW

File: packages/medusa/src/models/store.ts
Comment (srindom) [inline packages/medusa/src/models/store.ts:58]: thought (non-blocking): maybe we should also have a default location on the sales channel level. Nothing actionable right now, but worth keeping an eye out for if we start seeing a need for fallback behavior at the sales channel level.

File: packages/medusa/src/services/cart.ts
Comment (srindom) [inline packages/medusa/src/services/cart.ts:687]: nit: maybe could camelCase this following our discussion on casing in option/context objects

File: packages/medusa/src/services/cart.ts
Comment (srindom) [inline packages/medusa/src/services/cart.ts:799]: nit: maybe worth changing to `variantInventoryServiceTx`

File: packages/medusa/src/services/cart.ts
Comment (srindom) [inline packages/medusa/src/services/cart.ts:929]: nit: remove

File: packages/medusa/src/services/cart.ts
Comment (srindom) [inline packages/medusa/src/services/cart.ts:938]: nit: rm

File: packages/medusa/src/services/cart.ts
Comment (srindom) [inline packages/medusa/src/services/cart.ts:949]: nit: rm

File: packages/medusa/src/services/cart.ts
Comment (srindom) [inline packages/medusa/src/services/cart.ts:2071]: nit: rm

File: packages/medusa/src/services/product-variant-inventory.ts
Comment (srindom) [inline packages/medusa/src/services/product-variant-inventory.ts:1]: todo (this-file): please add JSDocs

File: packages/medusa/src/services/sales-channel-location.ts
Comment (srindom) [inline packages/medusa/src/services/sales-channel-location.ts:1]: todo (this-file): add JSDoc

File: packages/medusa/src/services/order.ts
Comment (srindom) [inline packages/medusa/src/services/order.ts:702]: thought (non-blocking): maybe it would make sense to move this into the cart completion strategy. My thinking is that consumers of the order service may want to have different workflows depending on their needs so decoupling the "reservation" step from the "placement" step would give more optionality. Nothing to do right now, but may make sense to include when we need to make the fulfillments location aware.

File: packages/medusa/src/strategies/cart-completion.ts
Comment (srindom) [inline packages/medusa/src/strategies/cart-completion.ts:17]: thought (this-file): this flow is great for @carlos-r-l-rodrigues' transaction orchestrator.

**Review by srindom — COMMENTED** [review]
Great work! Mostly looks good had a few nits and thoughts that are non blocking - otherwise just a bit of docs.

File: packages/medusa/src/models/store.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/models/store.ts:58]: @srindom, I will handle all these `location_id` fields on https://github.com/medusajs/medusa/pull/2907 

@pKorsholm I think we just have to align the migrations. You could have this field in a migration and I have split the Stock location and inventory in other 2 different migrations. If you add this field in this PR I can remove it from mine.

File: packages/medusa/src/migrations/1671711415179-multi_location.ts
Comment (olivermrbl) [inline packages/medusa/src/migrations/1671711415179-multi_location.ts:50]: todo: Remove public schema
```suggestion
    await queryRunner.query(
      `DROP INDEX "IDX_bf5386e7f2acc460adbf96d6f3"`
    )
    await queryRunner.query(
      `DROP INDEX "IDX_c74e8c2835094a37dead376a3b"`
    )
    await queryRunner.query(`DROP TABLE "product_variant_inventory_item"`)
    await queryRunner.query(
      `DROP INDEX "IDX_c2203162ca946a71aeb98390b0"`
    )
    await queryRunner.query(
      `DROP INDEX "IDX_6caaa358f12ed0b846f00e2dcd"`
    )
```

Comment (olivermrbl) [inline packages/medusa/src/migrations/1671711415179-multi_location.ts:50]: todo: Remove public schema
```suggestion
    await queryRunner.query(
      `DROP INDEX "IDX_bf5386e7f2acc460adbf96d6f3"`
    )
    await queryRunner.query(
      `DROP INDEX "IDX_c74e8c2835094a37dead376a3b"`
    )
    await queryRunner.query(`DROP TABLE "product_variant_inventory_item"`)
    await queryRunner.query(
      `DROP INDEX "IDX_c2203162ca946a71aeb98390b0"`
    )
    await queryRunner.query(
      `DROP INDEX "IDX_6caaa358f12ed0b846f00e2dcd"`
    )
```

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts:638]: suggestion(non-blocking): `if (lineItem.variant_id)`

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts:643]: suggestion(non-blocking): Extract to const before the statement for readability

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts:768]: suggestion(non-blocking): `if (item.variant_id)`

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts]: todo: Use camelCase

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts]: todo: Use camelCase

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts:2089]: todo: Add missing await
```suggestion
                return await lineItemServiceTx.update(item.id, {
```

File: packages/medusa/src/services/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/services/cart.ts:2095]: suggestion(non-blocking): 
```suggestion
                return await lineItemServiceTx.delete(item.id)
```

File: packages/medusa/src/services/claim-item.ts
Comment (olivermrbl) [inline packages/medusa/src/services/claim-item.ts:76]: question: Is virtual the correct word to use here? Would maybe choose `custom` or `Cannot claim line item that has no variant`

File: packages/medusa/src/services/product-variant-inventory.ts
Comment (olivermrbl) [inline packages/medusa/src/services/product-variant-inventory.ts:251]: todo: Extract to a type with camelCased props

File: packages/medusa/src/services/claim.ts
Comment (olivermrbl) [inline packages/medusa/src/services/claim.ts]: comment: See comment in `ProductVariantService`. Should be camelCased

File: packages/medusa/src/services/order.ts
Comment (olivermrbl) [reply packages/medusa/src/services/order.ts:702]: Really like this idea!

File: packages/medusa/src/services/order.ts
Comment (olivermrbl) [inline packages/medusa/src/services/order.ts]: comment: See comment in `ProductVariantService`. Should be camelCased

File: packages/medusa/src/services/product-variant-inventory.ts
Comment (olivermrbl) [inline packages/medusa/src/services/product-variant-inventory.ts:28]: todo: Should be suffixed with `_`
```suggestion
  protected readonly salesChannelLocationService_: SalesChannelLocationService
  protected readonly productVariantService_: ProductVariantService
  protected readonly stockLocationService_: IStockLocationService
  protected readonly inventoryService_: IInventoryService
```

File: packages/medusa/src/services/product-variant-inventory.ts
Comment (olivermrbl) [inline packages/medusa/src/services/product-variant-inventory.ts]: suggestion: To be consistent, we could rename this to `context`. Also, should be camelCased.

File: packages/medusa/src/services/product-variant-inventory.ts
Comment (olivermrbl) [inline packages/medusa/src/services/product-variant-inventory.ts:268]: todo:
```suggestion
      return await variantServiceTx.update(variant.id, {
        inventory_quantity: variant.inventory_quantity - quantity,
      })
```

File: packages/medusa/src/services/product-variant-inventory.ts
Comment (olivermrbl) [inline packages/medusa/src/services/product-variant-inventory.ts]: todo: Use `!isDefined(locationId)`

File: packages/medusa/src/services/sales-channel-inventory.ts
Comment (olivermrbl) [inline packages/medusa/src/services/sales-channel-inventory.ts:19]: todo: Should be suffixed with `_`
```suggestion
  protected readonly salesChannelLocationService_: SalesChannelLocationService
  protected readonly eventBusService_: EventBusService
  protected readonly inventoryService_: IInventoryService
```

File: packages/medusa/src/services/sales-channel-inventory.ts
Comment (olivermrbl) [inline packages/medusa/src/services/sales-channel-inventory.ts:15]:
```suggestion
  protected manager_: EntityManager
```

File: packages/medusa/src/services/sales-channel-location.ts
Comment (olivermrbl) [inline packages/medusa/src/services/sales-channel-location.ts:20]: todo:
```suggestion
  protected readonly salesChannelService_: SalesChannelService
  protected readonly eventBusService_: EventBusService
  protected readonly stockLocationService_: IStockLocationService
```

File: packages/medusa/src/strategies/cart-completion.ts
Comment (olivermrbl) [inline packages/medusa/src/strategies/cart-completion.ts:272]: todo: 
```suggestion
      const swap = await this.swapServiceTx_.retrieveByCartId(id)
```

**Review by olivermrbl — CHANGES_REQUESTED** [review]
Great piece of work! Have added a couple of suggestions and minor todos 💪

File: packages/medusa/src/services/order.ts
Comment (pKorsholm) [reply packages/medusa/src/services/order.ts]: I think this makes a lot of sense, I have moved this into the cart completion strategy 😄

File: packages/medusa/src/services/product-variant-inventory.ts
Comment (pKorsholm) [reply packages/medusa/src/services/product-variant-inventory.ts:309]: This is not allowed due to the return type of the method being void and we need to terminate early in this case. Alternative is to wrap everything following this in an else statement. I think this is better though

**Review by olivermrbl — COMMENTED** [review]
LGTM - great work! Pending @srindom's approval :)

**Review by olivermrbl — APPROVED** [review]

File: packages/medusa/src/strategies/cart-completion.ts
Comment (srindom) [inline packages/medusa/src/strategies/cart-completion.ts:310]: thought: this should probably go into its own step -- but let's not do anything, as we will refactor to the transaction orchestrator in any case.

**Review by srindom — APPROVED** [review]
LGTM w. comments - also think @carlos-r-l-rodrigues' suggestions are worth considering including

File: packages/medusa/src/strategies/cart-completion.ts
Comment (pKorsholm) [reply packages/medusa/src/strategies/cart-completion.ts:310]: 👍

**Review by carlos-r-l-rodrigues — APPROVED** [review]

**Comment by olivermrbl** [comment]
Ready to merge? 💪

**Pull request merged**
